### PR TITLE
US#1778031: better support for multibranch pipelines

### PR DESCRIPTION
**Pull request by dmytro-omelian**
Please Make sure these boxes are checked before submitting your pull request - Thanks ahead!

- [x] Proper pull request title - concise and clear for others.
- [ ] Proper pull request short description - clear and concise (as it should appear in the Jira ticket) for other developers and users.
- [ ] Proper Jira ticket - Number, Link in pull request description.
- [x] The PR can is merged on your machine without any conflicts.
- [x] The PR can is built on your machine without any (new) warnings.
- [ ] The PR passed sanity tests by you / QA / DevTest / Good Samaritan.
- [ ] Add unit tests with new features.
- [ ] If you added any dependency to the POM - Please update grount

### CODE REVIEW

File: HpToolsLauncher/Launcher.cs [hunk 1]
```diff
@@ -110,6 +111,11 @@ public static string DateFormat
             get { return _dateFormat; }
         }
 
+        public static string DefaultBranches
+        {
+            get { return _defaultBranches; }
+        }
+
```
Comment (nissimshitrit) [inline HpToolsLauncher/Launcher.cs]: is this part is necessary, this file is relevant to UFT test executions? can you please check your code is working if you remove your code from here?

File: src/main/java/com/microfocus/application/automation/tools/octane/CIJenkinsServicesImpl.java [hunk 2]
```diff
@@ -101,497 +104,537 @@
...
+    }
+
+    @Override
+    public CIBranchesList getBranchesList(String jobCiId, String filterBranchName) {
+        ACLContext securityContext = startImpersonation();
+        List<Branch> result = new ArrayList<>();
+        try {
+            Item item = Jenkins.get().getItemByFullName(jobCiId);
+            if (item != null) {
+                boolean hasAbortPermissions = item.hasPermission(Item.CANCEL);
```
Comment (nissimshitrit) [inline src/main/java/com/microfocus/application/automation/tools/octane/CIJenkinsServicesImpl.java]: why to you need to check if user have cancel permission? I think it relevant only for stop pipeline run action

File: src/main/java/com/microfocus/application/automation/tools/octane/CIJenkinsServicesImpl.java [hunk 3]
```diff
@@ -604,232 +647,257 @@ private PipelineNode createPipelineNode(String name, Job job, boolean includePar
...
+                .map(String::trim)
+                .filter(StringUtils::isNotEmpty)
+                .collect(Collectors.toSet());
+
+        Item item = Jenkins.get().getItemByFullName(name);
+        Collection<? extends Job> allJobs = item.getAllJobs();
+
+        Job job = allJobs.stream()
+                .filter(tempJob -> defaultBranches.contains(getDisplayNameFromJob(tempJob)))
+                .findFirst().orElse(null);
```
Comment (nissimshitrit) [inline src/main/java/com/microfocus/application/automation/tools/octane/CIJenkinsServicesImpl.java]: are you doing this for all jobs? can you do it only of multibranch jobs?

File: src/main/java/com/microfocus/application/automation/tools/octane/CIJenkinsServicesImpl.java [hunk 4]
```diff
@@ -604,232 +647,257 @@ private PipelineNode createPipelineNode(String name, Job job, boolean includePar
...
+
+        PipelineNode result = dtoFactory.newDTO(PipelineNode.class)
+                .setJobCiId(BuildHandlerUtils.translateFolderJobName(name))
+                .setName(name);
+
+        if(job != null) {
+            String defaultBranch = getDisplayNameFromJob(job);
+            result.setParameters(ParameterProcessors.getConfigs(job))
```
Comment (nissimshitrit) [inline src/main/java/com/microfocus/application/automation/tools/octane/CIJenkinsServicesImpl.java]: why do need to add the parameters? I didn't see on the original code?

**Review by nissimshitrit — CHANGES_REQUESTED** [review]

File: src/main/java/com/microfocus/application/automation/tools/octane/CIJenkinsServicesImpl.java [hunk 5]
```diff
@@ -604,232 +647,257 @@ private PipelineNode createPipelineNode(String name, Job job, boolean includePar
...
-		if (jobId != null && jobId.contains(BuildHandlerUtils.JOB_LEVEL_SEPARATOR)) {
-			int index = jobId.lastIndexOf(BuildHandlerUtils.JOB_LEVEL_SEPARATOR);
-			String parent = jobId.substring(0, index);
-			return parent;
-		}
-		return null;
-	}
-}
+    private PipelineNode createPipelineNodeFromJobName(String name) {
+        String[] defaultBranchesArray = config.getDefaultBranches().split(DEFAULT_BRANCHES_SEPARATOR);
```
Comment (nissimshitrit) [inline src/main/java/com/microfocus/application/automation/tools/octane/CIJenkinsServicesImpl.java]: can you also add NULL check here. it be null on some Jenkins configurations

**Review by nissimshitrit — CHANGES_REQUESTED** [review]

File: src/main/java/com/microfocus/application/automation/tools/octane/CIJenkinsServicesImpl.java (hunk 4 above)
Comment (dmytro-omelian) [reply src/main/java/com/microfocus/application/automation/tools/octane/CIJenkinsServicesImpl.java]: Actually, we do all those changes to get parameters from the default branch. I moved the logic for getting parameters to another method and it is only for multibranch jobs.

**Review by nissimshitrit — APPROVED** [review]

**Pull request merged**